tools/k-rule-find/main.cpp

```diff
@@ -0,0 +1,121 @@
+#include "kllvm/ast/AST.h"
+#include "kllvm/parser/KOREParser.h"
+#include "llvm/Support/CommandLine.h"
+
+#include <dlfcn.h>
+#include <iostream>
+#include <stdlib.h>
+
+#include "runtime/header.h"
+
+using namespace llvm;
+using namespace kllvm;
+
+struct Location {
+  char *filename;
+  int64_t line;
+  int64_t column;
+};
+
+std::string SOURCE_ATT = "org'Stop'kframework'Stop'attributes'Stop'Source";
+std::string LOCATION_ATT = "org'Stop'kframework'Stop'attributes'Stop'Location";
+
+cl::OptionCategory KRuleCat("k-rule-find options");
+
+cl::opt<std::string> KompiledDir(
+    cl::Positional, cl::desc("<kompiled-dir>"), cl::Required,
+    cl::cat(KRuleCat));
+
+cl::opt<std::string> RuleLocation(
+    cl::Positional, cl::desc("<filename.k:line[:column]>"), cl::Required,
+    cl::cat(KRuleCat));
+
+Location parseLocation(std::string loc) {
+  char *filename;
+  int64_t line, column = -1;
+  size_t pos = loc.find(":");
+  if (pos == std::string::npos) {
+    std::cerr
+        << "Rule's location must me in the format: defintion.k:line[:column]\n";
+    exit(EXIT_FAILURE);
+  }
+  filename = (char *)loc.substr(0, pos).c_str();
+
+  std::string lineColumn = loc.substr(pos + 1);
+  size_t pos_lc = lineColumn.find(":");
+
+  // If another “:” isn’t found, the tool assumes no column number was given.
+  if (pos_lc == std::string::npos) {
+    line = stoi(lineColumn);
+  } else {
+    line = stoi(lineColumn.substr(0, pos_lc));
+    column = stoi(lineColumn.substr(pos_lc + 1));
+  }
+
+  return {filename, line, column};
+}
+
+std::string getSource(KOREAxiomDeclaration *axiom) {
+  auto *sourceAtt = axiom->getAttributes().at(SOURCE_ATT).get();
+  assert(sourceAtt->getArguments().size() == 1);
+
+  auto strPattern
+      = dynamic_cast<KOREStringPattern *>(sourceAtt->getArguments()[0].get());
+  return strPattern->getContents();
+}
+
+Location getLocation(KOREAxiomDeclaration *axiom) {
```

Comment:
You will need two location types - see comment below, but remember that the locations in the definition are _ranges_. You need to extract their start and end lines / columns. The format is:
```
Location(start_line, start_col, end_line, end_col)
```

Reply:
Would you prefer a second Location type instead of extending the current one with `start_line`, `end_line`, etc.?
In this case, the input location will have `start_line == end_line`

Reply:
Yeah, that makes sense as an approach. The check you're implementing can then be a general one for whether one location range contains another